Add best practices resolving configurations eagerly (#32964)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/best-practices/bp_tasks.adoc

=== Tags
`#tasks`, `#caching`

== Don't resolve Configurations before Task Execution

Resolving configuration before task execution phase can lead to incorrect results and slows gradle build performance.

> Once configuration is resolved it can't be resolved again, and it can lose some information when it is resolved too early.

Reviewer: I think this needs more detail.

Author: When I wrote this, I thought about two potential problems:
* task dependencies loss
* items added/modified to/in configuration after resolution

But the second part is very-well guarded by Gradle's runtime. it reports errors and warnings when something fishy is happening with Configuration.  So perhaps I can cut this part and focus only on the task dependencies, as this info is not tracked by Gradle runtime (yet).

> Calling manual resolution via link:{javadocPath}/org/gradle/api/artifacts/Configuration.html#resolve()[resolve] method returns a set of Files that doesn't preserve references to the tasks that produced these files.

Reviewer: I think the potential for losing dependency info should be a separate paragraph.

Author: I separated them, but didn't fall into thorough explanation. Does it look good?
I'd probably stick to short explanations with links if available.
By the way, do you know any good links to other gradle docs on this topic that could fit here?

Reviewer: @lkasso ?

Author: Since @lkasso approved the PR, maybe we can let it go as is? And correct if necessary later.
If you're ok with the current explanation.

This will break automatic task dependencies wiring through task input properties.

=== Example

==== Don't Do This

====
include::sample[dir="snippets/bestPractices/avoidResolvingConfigurationsManually/kotlin",files="build.gradle.kts[tags=resolve-call]"]
include::sample[dir="snippets/bestPractices/avoidResolvingConfigurationsManually/groovy",files="build.gradle[tags=resolve-call]"]
====

<1> *Explicit resolve call*: `runtimeClasspath` configuration will be resolved as soon as task created. It will lose all task dependencies as well.

==== Do This Instead

Use lazy link:{javadocPath}/org/gradle/api/file/FileCollection.html[FileCollection] API instead.

====
include::sample[dir="snippets/bestPractices/avoidResolvingConfigurationsManually/kotlin",files="build.gradle.kts[tags=good-classpath-printer]"]
include::sample[dir="snippets/bestPractices/avoidResolvingConfigurationsManually/groovy",files="build.gradle[tags=good-classpath-printer]"]
====

> <1> *Pass runtimeClasspath as input to the File Collection*: the File Collection will not resolve `runtimeClasspath` immediately, and will preserve its task dependencies.

Reviewer: I think the preserving task dependencies is an important part of this.  In this example there are no task dependencies to preserve, so noting that it will preserve task dependencies here is odd.

Maybe this example should focus on the other negative impacts of configuration-time resolution.  Or should include a task dependency (with the extra complexity that entails) - then it could demo the resolution properly causing the task to run and improperly not doing so if resolved at configuration time.

Author: I've reworked this sample entirely. Now I add separate sub-project and declare dependency on it.

> <2> *Access resolved files*: It is safe to access files from the File Collection during task execution.

Reviewer: Is this meant to imply accessing files is _not_ safe during configuration time?

Author: I've reworked the sample entirely. Please re-check new version.

=== Tags
`#tasks`, `#inputs-and-outputs`, `#configurations`

> == Avoid using eager API on Gradle File Collections before task execution

Reviewer: We should try to keep to 1 item per PR.

Author: Ok!

Gradle link:{javadocPath}/org/gradle/api/artifacts/Configuration.html[Configuration] and link:{javadocPath}/org/gradle/api/artifacts/FileCollection.html[FileCollection]

> extend JDK's `Collection<File>` interface, however calls to the JDK Collection API on configuration and file collection instances will trigger their implicit resolution.

Reviewer: I think we could broaden this advice to also avoid the eager APIs Gradle adds as well.

Author: I'm not sure what other API should be mention here then? Do you have something in mind?

Reviewer: This is good now that the item is limited to file collections.

=== Example
==== Don't Do This

====
include::sample[dir="snippets/bestPractices/avoidJdkCollectionApiOnFileCollections/kotlin",files="build.gradle.kts[tags=bad-zipping-task]"]
include::sample[dir="snippets/bestPractices/avoidJdkCollectionApiOnFileCollections/groovy",files="build.gradle[tags=bad-zipping-task]"]
====

> <1> *Add file via plus operator*: Plus operator will force runtimeClasspath configuration to be resolved implicitly. Task dependencies will be lost.

Reviewer: I think this needs a mention of _why_ the plus operator does this.

Author: Do you mean the fact that Configuration or FileCollection don't override plus operator?

Author: Updated.

- <<build_cache_concepts.adoc#non_cacheable_tasks,Non-cacheable Tasks>>

=== Tags

> `#tasks`, `#caching`

Reviewer 2: Let's try not to override this.

== Don't resolve Configurations before Task Execution

> Resolving configuration before the task execution phase can lead to incorrect results and slower builds.

Reviewer 2: ```suggestion
Resolving configurations before the task execution phase can lead to incorrect results and slower builds.
```

=== Explanation
 

> Calling manual resolution via link:{javadocPath}/org/gradle/api/artifacts/Configuration.html#resolve()[resolve] method returns a set of Files that doesn't preserve references to the tasks that produced these files.

Reviewer 2: ```suggestion
Calling manual resolution via the link:{javadocPath}/org/gradle/api/artifacts/Configuration.html#resolve()[`resolve()`] method returns a set of files that does not preserve references to the tasks that produced those files.
```

> This will break automatic task dependencies wiring through task input properties.

Reviewer 2: ```suggestion
This breaks automatic task dependency wiring through task input properties.
```

> Resolving configuration outside task execution phase can slow down configuration phase even for tasks (e.g. `help`) that don't need results of the resolved configuration.

Reviewer 2: ```suggestion
In addition, resolving configurations during the configuration phase can slow down the build, even for tasks (e.g., `help`) that don't require the resolved dependencies.
```

=== Example

==== Don't Do This

====
include::sample[dir="snippets/bestPractices/avoidResolvingConfigurationsManually/kotlin",files="library/build.gradle.kts;build.gradle.kts[tags=bad-classpath-printer]"]
include::sample[dir="snippets/bestPractices/avoidResolvingConfigurationsManually/groovy",files="library/build.gradle;build.gradle[tags=bad-classpath-printer]"]
====

<1> *Add project dependency*: It implies that `:library:jar` task should be executed in order to resolve runtime classpath.
<2> *Declare input files property as Set of files*: It doesn't track task dependencies.
<3> *Pass resolved configuration to input files property*: Dependency on `:library:jar` task will be lost here.
<4> *Consume actual jar file or classes directory to calculate digest*: Due to broken task dependencies, library.jar file will not exist at this point.

==== Do This Instead

> Use lazy link:{javadocPath}/org/gradle/api/file/FileCollection.html[FileCollection] API instead.

Reviewer 2: ```suggestion
To avoid issues, always defer resolution to the execution phase by using lazy APIs like link:{javadocPath}/org/gradle/api/file/FileCollection.html[`FileCollection`].
```

====
include::sample[dir="snippets/bestPractices/avoidResolvingConfigurationsManually/kotlin",files="library/build.gradle.kts;build.gradle.kts[tags=good-classpath-printer]"]
include::sample[dir="snippets/bestPractices/avoidResolvingConfigurationsManually/groovy",files="library/build.gradle;build.gradle[tags=good-classpath-printer]"]
====

<1> *Add project dependency*: It implies that jar task should be executed in order to resolve runtime classpath.
<2> *Declare input files property as ConfigurableFileCollection*: It will track task dependencies.
<3> *Pass configuration as-is to input files property*: It will be resolved when necessary and task dependencies will be preserved.
<4> *Consume actual jar file or classes directory to calculate digest*: Here all files should exist at this point.

=== Tags

> `#tasks`, `#inputs-and-outputs`, `#configurations`

Reviewer 2: Do you mind updating this to linkable versions: `<<tags_reference.adoc#tag:tasks,#tasks>>`?

Author: Sure, I'll update it accordingly

> == Avoid using eager API on Gradle File Collections before task execution

Reviewer 2: ```suggestion
== Avoid using eager APIs on File Collections before Task Execution
```

> Gradle link:{javadocPath}/org/gradle/api/artifacts/Configuration.html[Configuration] and link:{javadocPath}/org/gradle/api/artifacts/FileCollection.html[FileCollection]

Reviewer 2: ```suggestion
Gradle's link:{javadocPath}/org/gradle/api/artifacts/Configuration.html[`Configuration`] and link:{javadocPath}/org/gradle/api/file/FileCollection.html[`FileCollection`] types extend the JDK’s `Collection<File>` interface.
```

> extend JDK's `Collection<File>` interface, however calls to the JDK Collection API on configuration and file collection instances will trigger their implicit resolution.

Reviewer 2: ```suggestion
However, calling eager methods—such as `.size()`, `.isEmpty()`, `.files`, `.asPath`, or `.toList()`—on these types during the configuration phase will trigger their implicit resolution.
```

=== Example
==== Don't Do This

====
include::sample[dir="snippets/bestPractices/avoidJdkCollectionApiOnFileCollections/kotlin",files="build.gradle.kts[tags=bad-zipping-task]"]
include::sample[dir="snippets/bestPractices/avoidJdkCollectionApiOnFileCollections/groovy",files="build.gradle[tags=bad-zipping-task]"]
====

<1> *Add file via plus operator*: Plus operator will force runtimeClasspath configuration to be resolved implicitly. Configuration implementation doesn't override plus operator for regular files, therefore it fallsback to eager API. Task dependencies will be lost here.

==== Do This Instead

> Use Gradle API that supports lazy evaluation.

Reviewer 2: ```suggestion
To avoid issues, always defer resolution until the execution phase. 
Use APIs that support lazy evaluation.
```

====
include::sample[dir="snippets/bestPractices/avoidJdkCollectionApiOnFileCollections/kotlin",files="build.gradle.kts[tags=good-zipping-task]"]
include::sample[dir="snippets/bestPractices/avoidJdkCollectionApiOnFileCollections/groovy",files="build.gradle[tags=good-zipping-task]"]
====

<1> *Pass files separately*: This allows combining file collections without their implicit resolution.

=== Tags

> `#tasks`, `#inputs-and-outputs`, `#configurations`

Reviewer 2: Let's make these links
